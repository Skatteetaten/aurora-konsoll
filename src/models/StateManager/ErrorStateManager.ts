import StateManager from 'models/StateManager';

export interface IAppError {
  id: number;
  error: Error;
  isActive: boolean;
}

export interface IErrorState {
  errorQueue: IAppError[];
  allErrors: Map<number, IAppError>;
}

class ErrorStateManager extends StateManager<IErrorState> {
  public errorCount = 0;

  public addError(error: Error) {
    const state = this.getState();
    state.errorQueue.unshift({
      error,
      id: this.errorCount++,
      isActive: true
    });
    this.updateState(state);
  }

  public getNextError(): IAppError | undefined {
    const state = this.getState();
    const next = state.errorQueue.pop();
    if (!next) {
      return;
    }
    state.allErrors.set(next.id, next);
    this.updateState(state);
    return next;
  }

  public closeError = (id: number) => {
    const state = this.getState();
    const err = state.allErrors.get(id);
    if (!err) {
      throw new Error('No such error ' + id);
    }
    err.isActive = false;
    this.updateState(state);
  };

  public hasError() {
    return this.getState().errorQueue.length > 0;
  }

  public registerStateUpdater(updater: (errors: IErrorState) => void) {
    this.onUpdateState = updater;
  }
}

const errorSM = new ErrorStateManager(
  { errorQueue: [], allErrors: new Map() },
  () => {
    return;
  }
);

export { errorSM };
export default ErrorStateManager;
