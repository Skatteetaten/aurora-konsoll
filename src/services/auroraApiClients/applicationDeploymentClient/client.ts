import {
  IApplicationDeployment,
  IApplicationDeploymentDetails,
  IUserAndAffiliations
} from 'models/ApplicationDeployment';
import { normalizeRawDeploymentSpec } from 'models/DeploymentSpec';

import { IPodResource } from 'models/Pod';
import GoboClient from 'services/GoboClient';
import {
  REDEPLOY_WITH_CURRENT_VERSION_MUTATION,
  REDEPLOY_WITH_VERSION_MUTATION,
  REFRESH_APPLICATION_DEPLOYMENT_MUTATION,
  REFRESH_APPLICATION_DEPLOYMENTS_MUTATION
} from './mutation';
import {
  APPLICATION_DEPLOYMENT_DETAILS_QUERY,
  APPLICATIONS_QUERY,
  IApplicationDeploymentDetailsQuery,
  IApplicationsConnectionQuery,
  IUserAffiliationsQuery,
  USER_AFFILIATIONS_QUERY
} from './query';

function formatName(user: string) {
  const names = user.split(', ');
  if (names.length !== 2) {
    return user;
  }
  return names[1] + ' ' + names[0];
}

export class ApplicationDeploymentClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async redeployWithVersion(
    applicationDeploymentId: string,
    version: string
  ): Promise<boolean> {
    const result = await this.client.mutate<{ redeployWithVersion: boolean }>({
      mutation: REDEPLOY_WITH_VERSION_MUTATION,
      variables: {
        input: {
          applicationDeploymentId,
          version
        }
      }
    });

    if (result && result.data) {
      return result.data.redeployWithVersion;
    }

    return false;
  }

  public async redeployWithCurrentVersion(
    applicationDeploymentId: string
  ): Promise<boolean> {
    const result = await this.client.mutate<{
      redeployWithCurrentVersion: boolean;
    }>({
      mutation: REDEPLOY_WITH_CURRENT_VERSION_MUTATION,
      variables: {
        input: {
          applicationDeploymentId
        }
      }
    });

    if (result && result.data) {
      return result.data.redeployWithCurrentVersion;
    }

    return false;
  }

  public async refreshApplicationDeployment(
    applicationDeploymentId: string
  ): Promise<boolean> {
    const result = await this.client.mutate<{
      refreshApplicationDeployment: string;
    }>({
      mutation: REFRESH_APPLICATION_DEPLOYMENT_MUTATION,
      variables: {
        input: {
          applicationDeploymentId
        }
      }
    });

    if (result && result.data) {
      return true;
    } else {
      return false;
    }
  }

  public async refreshAffiliations(affiliations: string[]): Promise<boolean> {
    const result = await this.client.mutate<{
      affiliations: string[];
    }>({
      mutation: REFRESH_APPLICATION_DEPLOYMENTS_MUTATION,
      variables: {
        input: {
          affiliations
        }
      }
    });

    if (result && result.data) {
      return true;
    } else {
      return false;
    }
  }

  public async findApplicationDeploymentDetails(
    id: string
  ): Promise<IApplicationDeploymentDetails> {
    const result = await this.client.query<IApplicationDeploymentDetailsQuery>({
      query: APPLICATION_DEPLOYMENT_DETAILS_QUERY,
      variables: {
        id
      }
    });
    let deploymentSpec;
    let pods: IPodResource[] = [];
    if (result && result.data && result.data.applicationDeploymentDetails) {
      const {
        deploymentSpecs,
        podResources
      } = result.data.applicationDeploymentDetails;

      if (deploymentSpecs.current) {
        const spec = JSON.parse(deploymentSpecs.current.jsonRepresentation);
        deploymentSpec = Object.keys(spec).reduce(
          normalizeRawDeploymentSpec(spec),
          {}
        );
      }
      pods = podResources;
    }

    return {
      deploymentSpec,
      pods
    };
  }

  public async findUserAndAffiliations(): Promise<IUserAndAffiliations> {
    const result = await this.client.query<IUserAffiliationsQuery>({
      query: USER_AFFILIATIONS_QUERY
    });

    if (!result) {
      return {
        affiliations: [],
        user: ''
      };
    }

    return {
      affiliations: result.data.affiliations.edges.map(edge => edge.node.name),
      user: formatName(result.data.currentUser.name)
    };
  }

  public async findAllApplicationDeployments(
    affiliations: string[]
  ): Promise<IApplicationDeployment[]> {
    const result = await this.client.query<IApplicationsConnectionQuery>({
      query: APPLICATIONS_QUERY,
      variables: {
        affiliations
      }
    });

    if (!result) {
      return [];
    }

    return result.data.applications.edges.reduce((acc, { node }) => {
      const { applicationDeployments, imageRepository } = node;
      const deployments = applicationDeployments.map(app => ({
        affiliation: app.affiliation.name,
        environment: app.environment,
        id: app.id,
        name: app.name,
        permission: app.namespace.permission,
        repository: imageRepository ? imageRepository.repository : '',
        status: {
          code: app.status.code,
          reasons: app.status.reasons,
          reports: app.status.reports
        },
        time: app.time,
        version: {
          releaseTo: app.version.releaseTo,
          auroraVersion: app.version.auroraVersion,
          deployTag: {
            lastModified: '',
            name: findDeployTagForTemplate(
              node.name,
              app.version.deployTag.name
            ),
            type: app.version.deployTag.type
          }
        }
      }));
      return [...acc, ...deployments];
    }, []);
  }
}

// ! Temp fix for template deployments with default version
// TODO: FIX
function findDeployTagForTemplate(applicationName: string, deployTag: string) {
  const templates = {
    'aurora-activemq-1.0.0': '2',
    'aurora-redis-1.0.0': '3.2.3',
    'aurora-wiremock-1.0.0': '1.3.0',
    redis: '3.2.3',
    wiremock: '1.3.0'
  };

  if (deployTag) {
    return deployTag;
  }

  return templates[applicationName] || deployTag;
}
