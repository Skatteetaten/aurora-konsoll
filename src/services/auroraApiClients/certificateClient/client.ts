import { ICertificateResult } from 'models/certificates';
import GoboClient from 'services/GoboClient';
import { CERTIFICATES_QUERY, ICertificate, ICertificatesQuery } from './query';

export class CertificateClient {
  public defaultCertificates: ICertificatesQuery = {
    certificates: {
      totalCount: 0,
      edges: []
    }
  };

  public client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }
  public async getCertificates(): Promise<ICertificateResult> {
    const result = await this.client.query<ICertificatesQuery>({
      query: CERTIFICATES_QUERY
    });
    if (!result) {
      return {
        certificates: [],
        totalCount: 0
      };
    }

    return {
      certificates: this.normalizeScanStatus(result.data),
      totalCount: result.data.certificates.totalCount
    };
  }

  private normalizeScanStatus(data: ICertificatesQuery): ICertificate[] {
    if (!data) {
      return [];
    }

    return data.certificates.edges.map(edge => {
      const { cn, expiresDate, id, issuedDate, revokedDate } = edge.node;
      return {
        issuedDate,
        cn,
        expiresDate,
        id,
        revokedDate
      };
    });
  }
}
