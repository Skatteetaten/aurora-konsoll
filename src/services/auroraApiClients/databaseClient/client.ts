import GoboClient, { IDataAndErrors } from 'services/GoboClient';

import {
  ICreateDatabaseSchemaInput,
  IDeleteDatabaseSchemasResponse,
  IJdbcUser,
  IUpdateDatabaseSchemaInputWithCreatedBy
} from 'models/schemas';
import {
  CREATE_DATABASE_SCHEMA_MUTATION,
  DELETE_DATABASESCHEMAS_MUTATION,
  TEST_JDBC_CONNECTION_FOR_ID_MUTATION,
  TEST_JDBC_CONNECTION_FOR_JDBCUSER_MUTATION,
  UPDATE_DATABASESCHEMA_MUTATION
} from './mutation';
import {
  DATABASE_SCHEMAS_QUERY,
  IDatabaseSchemasQuery,
  DATABASE_INSTANCES_QUERY,
  IDatabaseInstancesQuery
} from './query';

export class DatabaseClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getSchemas(
    affiliations: string[]
  ): Promise<IDataAndErrors<IDatabaseSchemasQuery> | undefined> {
    return await this.client.query<IDatabaseSchemasQuery>({
      query: DATABASE_SCHEMAS_QUERY,
      variables: {
        affiliations
      }
    });
  }

  public async getInstances(
    affiliation: string
  ): Promise<IDataAndErrors<IDatabaseInstancesQuery> | undefined> {
    return await this.client.query<IDatabaseInstancesQuery>({
      query: DATABASE_INSTANCES_QUERY,
      variables: {
        affiliation
      }
    });
  }

  public async updateSchema(
    databaseSchema: IUpdateDatabaseSchemaInputWithCreatedBy
  ): Promise<
    IDataAndErrors<{ updateDatabaseSchema: { id: string } }> | undefined
  > {
    return await this.client.mutate<{ updateDatabaseSchema: { id: string } }>({
      mutation: UPDATE_DATABASESCHEMA_MUTATION,
      variables: {
        input: databaseSchema
      }
    });
  }

  public async deleteSchemas(
    ids: string[]
  ): Promise<
    | IDataAndErrors<{
        deleteDatabaseSchemas: IDeleteDatabaseSchemasResponse;
      }>
    | undefined
  > {
    return await this.client.mutate<{
      deleteDatabaseSchemas: IDeleteDatabaseSchemasResponse;
    }>({
      mutation: DELETE_DATABASESCHEMAS_MUTATION,
      variables: {
        input: {
          ids
        }
      }
    });
  }

  public async testJdbcConnectionForId(
    id: string
  ): Promise<
    | IDataAndErrors<{
        testJdbcConnectionForId: boolean;
      }>
    | undefined
  > {
    return await this.client.mutate<{ testJdbcConnectionForId: boolean }>({
      mutation: TEST_JDBC_CONNECTION_FOR_ID_MUTATION,
      variables: {
        id
      }
    });
  }

  public async testJdbcConnectionForJdbcUser(jdbcUser: IJdbcUser) {
    return await this.client.mutate<{
      testJdbcConnectionForJdbcUser: boolean;
    }>({
      mutation: TEST_JDBC_CONNECTION_FOR_JDBCUSER_MUTATION,
      variables: {
        input: jdbcUser
      }
    });
  }

  public async createDatabaseSchema(
    databaseSchema: ICreateDatabaseSchemaInput
  ) {
    return await this.client.mutate<{
      createDatabaseSchema: {
        id: string;
        jdbcUrl: string;
        users: [{ username: string; password: string }];
      };
    }>({
      mutation: CREATE_DATABASE_SCHEMA_MUTATION,
      variables: {
        input: databaseSchema
      }
    });
  }
}
