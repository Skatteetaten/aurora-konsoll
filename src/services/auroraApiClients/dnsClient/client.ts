import GoboClient, { IDataAndErrors } from 'services/GoboClient';
import { CnameInfosQuery, CNAME_INFO_QUERY } from './query';

export class DnsClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async fetchCnameInfos(
    affiliation: string
  ): Promise<IDataAndErrors<CnameInfosQuery>> {
    return await this.client.query<CnameInfosQuery>({
      query: CNAME_INFO_QUERY,
      variables: {
        affiliation,
      },
    });
  }
}
