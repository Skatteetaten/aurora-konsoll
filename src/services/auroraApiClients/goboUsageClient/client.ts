import GoboClient, { IGoboResult } from 'services/GoboClient';
import { GOBO_USERS_USAGE_QUERY, IGoboUsage } from './query';

export class GoboUsageClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getGoboUsers(): Promise<IGoboResult<IGoboUsage> | undefined> {
    return await this.client.query<IGoboUsage>({
      query: GOBO_USERS_USAGE_QUERY
    });
  }
}
