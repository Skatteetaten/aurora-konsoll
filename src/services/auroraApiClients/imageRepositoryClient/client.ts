import ApolloClient from 'apollo-boost';

import { ITagsPagedGroup, TagsPagedGroup } from 'services/auroraApiClients';

import {
  IImageTagsConnection,
  ITagsGroupedQuery,
  ITagsQuery,
  TAGS_GROUPED_QUERY,
  TAGS_QUERY
} from './query';

export interface ITag {
  name: string;
  lastModified: string;
}

export interface ITagsPaged {
  endCursor: string;
  hasNextPage: boolean;
  tags: ITag[];
}

export interface IImageRepositoryClient {
  findTagsPaged: (
    repository: string,
    first?: number,
    cursor?: string,
    types?: string[]
  ) => Promise<ITagsPaged>;
  findGroupedTagsPaged: (repository: string) => Promise<TagsPagedGroup>;
}

export class ImageRepositoryClient implements IImageRepositoryClient {
  private client: ApolloClient<{}>;

  constructor(client: ApolloClient<{}>) {
    this.client = client;
  }

  public async findTagsPaged(
    repository: string,
    first?: number,
    cursor?: string,
    types?: string[]
  ): Promise<ITagsPaged> {
    const result = await this.client.query<ITagsQuery>({
      query: TAGS_QUERY,
      variables: {
        cursor,
        first,
        repositories: [repository],
        types
      }
    });

    const { imageRepositories } = result.data;

    if (!(imageRepositories && imageRepositories.length > 0)) {
      throw new Error(`Could not find tags for repository ${repository}`);
    }

    return this.toTagsPaged(imageRepositories[0].tags);
  }

  public async findGroupedTagsPaged(
    repository: string
  ): Promise<TagsPagedGroup> {
    const result = await this.client.query<ITagsGroupedQuery>({
      query: TAGS_GROUPED_QUERY,
      variables: {
        repositories: [repository]
      }
    });

    const { imageRepositories } = result.data;

    if (!(imageRepositories && imageRepositories.length > 0)) {
      throw new Error(`Could not find tags for repository ${repository}`);
    }

    const [mainRepo] = imageRepositories;

    const normalizedTags: ITagsPagedGroup = {
      auroraVersion: this.toTagsPaged(mainRepo.auroraVersion),
      bugfix: this.toTagsPaged(mainRepo.bugfix),
      latest: this.toTagsPaged(mainRepo.latest),
      major: this.toTagsPaged(mainRepo.major),
      minor: this.toTagsPaged(mainRepo.minor),
      snapshot: this.toTagsPaged(mainRepo.snapshot)
    };

    return new TagsPagedGroup(normalizedTags);
  }
  private toTagsPaged(imageTagsConnection: IImageTagsConnection): ITagsPaged {
    const { edges, pageInfo } = imageTagsConnection;
    return {
      endCursor: pageInfo.endCursor,
      hasNextPage: pageInfo.hasNextPage,
      tags: edges.map(edge => ({
        lastModified: edge.node.lastModified,
        name: edge.node.name
      }))
    };
  }
}
