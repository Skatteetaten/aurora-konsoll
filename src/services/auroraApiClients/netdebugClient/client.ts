import ApolloClient from 'apollo-boost';

import { IScanQuery, IScanStatusQuery, NETDEBUG_QUERY } from './query';

export interface INetdebugResult {
  status: string;
  open: IScanStatus[];
  failed: IScanStatus[];
}
export interface IScanStatus {
  status: string;
  message?: string;
  clusterNodeIp?: number;
}

export class NetdebugClient {
  private client: ApolloClient<{}>;

  constructor(client: ApolloClient<{}>) {
    this.client = client;
  }

  public async findNetdebugStatus(
    host: string,
    port: string
  ): Promise<INetdebugResult> {
    const result = await this.client.query<IScanQuery>({
      query: NETDEBUG_QUERY,
      variables: {
        host,
        port
      }
    });
    return this.showNetdebugStatus(result.data);
  }

  private showNetdebugStatus(item: IScanQuery): INetdebugResult {
    return {
      failed: this.normalizeScanStatus(item.scan.failed),
      open: this.normalizeScanStatus(item.scan.open),
      status: item.scan.status
    };
  }

  private normalizeScanStatus(scanStatus?: IScanStatusQuery): IScanStatus[] {
    if (!scanStatus) {
      return [];
    }

    return scanStatus.edges.map(edge => {
      const { clusterNode, message, status } = edge.node;
      return {
        clusterNodeIp: clusterNode && clusterNode.ip,
        message,
        status
      };
    });
  }
}
