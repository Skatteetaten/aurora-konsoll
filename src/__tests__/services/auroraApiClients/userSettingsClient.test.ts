import ErrorStateManager from 'models/StateManager/ErrorStateManager';
import { UserSettingsClient } from 'services/auroraApiClients/userSettingsClient/client';
import { goboClientMock, GraphQLSeverMock } from 'utils/GraphQLMock';

import * as getUserSettings from './__responses__/userSettingsClient/getUserSettings.json';

const errorStateManager = new ErrorStateManager(
  {
    allErrors: new Map(),
    errorQueue: []
  },
  () => {
    // Validate errors
    return;
  }
);

const serverMock = new GraphQLSeverMock();
const clientMock = goboClientMock(serverMock.graphQLUrl, errorStateManager);
const userSettingsClient = new UserSettingsClient(clientMock);

afterAll(() => {
  serverMock.close();
});

describe('getUserSettings', () => {
  it('should fetch user settings from GraphQL server and normalize data', async () => {
    serverMock.putResponse('getUserSettingsForAllAffiliations', getUserSettings);

    const result = await userSettingsClient.getUserSettings();
    expect(result).toMatchSnapshot();
  });
});
