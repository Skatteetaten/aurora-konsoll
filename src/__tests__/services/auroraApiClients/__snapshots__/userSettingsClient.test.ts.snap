// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserSettings should fetch user settings from GraphQL server and normalize data 1`] = `
Object {
  "applicationDeploymentFilters": Array [
    Object {
      "affiliation": "aurora",
      "applications": Array [
        "ao",
      ],
      "default": true,
      "environments": Array [
        "dev",
      ],
      "name": "testing",
    },
  ],
}
`;

exports[`updateUserSettings should update user settings to GraphQL server 1`] = `true`;
